homeassistant/components/zha/core/helpers.py

```diff
@@ -403,13 +403,13 @@ def convert_install_code(value: str) -> bytes:
         [0-9a-fA-F]{34}
         ([0-9a-fA-F]{16}) # IEEE address
         DLK
-        ([0-9a-fA-F]{36}) # install code
+        ([0-9a-fA-F]{36}|[0-9a-fA-F]{32}) # install code / link key
         $
     """,
 )
 
 
-def qr_to_install_code(qr_code: str) -> tuple[zigpy.types.EUI64, bytes]:
+def qr_to_install_code(qr_code: str) -> tuple[zigpy.types.EUI64, bytes | zigpy.types.KeyData]:
```

Comment:
The preferred interface for permitting with install codes is currently just `permit_with_link_key`, with the link key derived within the caller. I'm hoping to drop support for `permit_with_key` at some point.

Perhaps it would be simpler to rename this method `def qr_to_link_key(qr_code: str) -> tuple[zigpy.types.EUI64, zigpy.types.KeyData]` and have the Bosch QR fall through as it does now, while returning a link key derived from the install code in all other cases? That way, this is the only spot it needs to be special cased.

This should also clean up the websocket API too, as we would only ever need to call `permit_with_link_key`.

Reply:
Very good point! I updated this PR do replace `permit_with_key` by `permit_with_link_key`. Now `qr_to_install_code` and `convert_install_code` will supply directly KeyData.